module/zfs/dmu_tx.c

```diff
@@ -1055,6 +1056,9 @@ dmu_tx_assign(dmu_tx_t *tx, uint64_t txg_how)
 
 	txg_rele_to_quiesce(&tx->tx_txgh);
 
+	if (dsl_pool_need_dirty_sync(tx->tx_pool, tx->tx_txg)) {
+		txg_kick(tx->tx_pool, tx->tx_txg);
+	}
```

Comment:
It seems like it might be somewhat expensive to do this check which takes and releases the `dp->dp_lock` for every assign.  Though since we take the same lock every time we dirty/undirty bytes this is probably fine.  Did you happen to look at the contention on this lock when pushing a large number of transactions?

nit cstyle: let's drop the optional parenthesis here to be consistent with the rest of the code.

```c
 	if (dsl_pool_need_dirty_sync(tx->tx_pool, tx->tx_txg))
 		txg_kick(tx->tx_pool, tx->tx_txg);

 	return (0);
```

Reply:
I haven't had chance to verify the performance implication of this commit. I plan to do some simple comparison test on ramdisk. 
Regarding the lock, the overhead might be mitigated with atomic operation instead of mutex. However, it will be another pull request.